qtfred/src/mission/dialogs/CampaignEditorDialogModel.cpp

```diff
@@ -0,0 +1,801 @@
+
+static QString loadFile(QString file, const QString& campaignType) {
+	if (file.isEmpty()) {
+		mission_campaign_clear();
+		Campaign.type = campaignType.isEmpty() ? CAMPAIGN_TYPE_SINGLE :
+												 CampaignEditorDialogModel::campaignTypes.indexOf(campaignType);
+		return {};
+	}
+	//FRED is to enforce that only on new campaigns a campaign type may be given
+	Assert(campaignType.isEmpty());
+	parse_init(false);
+	if (mission_campaign_load(qPrintable(file.replace('/',DIR_SEPARATOR_CHAR)), nullptr, 0))
+		return {};
+
+	return Campaign.filename;
+}
+
+static void initShips(const SCP_vector<ship_info>::const_iterator &s_it, CheckedDataListModel<std::ptrdiff_t> &model){
+	std::ptrdiff_t shpIdx{ std::distance(Ship_info.cbegin(), s_it) };
+	if (s_it->flags[Ship::Info_Flags::Player_ship]) {
+		model.initRow(
+				s_it->name,
+				shpIdx,
+				static_cast<bool>(Campaign.ships_allowed[static_cast<size_t>(shpIdx)]));
+}
+}
+
+static void initWeps(const SCP_vector<weapon_info>::const_iterator &w_it, CheckedDataListModel<std::ptrdiff_t> &model){
+	std::ptrdiff_t wepIdx{ std::distance(Weapon_info.cbegin(), w_it) };
+	for (const ship_info& shp: Ship_info) {
+		if (shp.flags[Ship::Info_Flags::Player_ship]
+			&& shp.allowed_weapons[static_cast<size_t>(wepIdx)]
+			&& !model.contains(w_it->name)) {
+			model.initRow(
+					w_it->name,
+					wepIdx,
+					static_cast<bool>(Campaign.weapons_allowed[static_cast<size_t>(wepIdx)]));
+}
+}
+}
+
+static SCP_vector<SCP_string> getMissions(){
+	Skip_packfile_search = 1;
+	SCP_vector<SCP_string> missions;
+	cf_get_file_list(missions, CF_TYPE_MISSIONS, "*fs2", CF_SORT_NAME);
+	Skip_packfile_search = 0;
+	return missions;
+}
+
+static bool parseMnPart(mission *mn, const char *filename){
+	try {
+		read_file_text(filename, CF_TYPE_MISSIONS);
+
+		parse_init(false);
+		Subsys_index = 0;
+		Subsys_status_size = 0;
+		if (Subsys_status != nullptr) {
+			vm_free( Subsys_status );
+			Subsys_status = nullptr;
+		}
+
+		parse_mission_info(mn, true);
+		skip_to_start_of_string("#Events");
+		Num_mission_events = 0;
+		parse_events(mn);
+		Num_goals = 0;
+		parse_goals(mn);
+	}  catch (const parse::ParseException& ) {
+		return false;
+	}
+	return true;
+}
+
+CampaignEditorDialogModel::CampaignEditorDialogModel(CampaignEditorDialog* _parent, EditorViewport* viewport, const QString &file, const QString& newCampaignType) :
+	AbstractDialogModel(_parent, viewport),
+	parent(_parent),
+	campaignFile(loadFile(file, newCampaignType)),
+	campaignType(campaignTypes[Campaign.type]),
+	initialShips(Ship_info, &initShips, this),
+	initialWeapons(Weapon_info, &initWeps, this),
+	missionData(getMissions(), &CampaignMissionData::initMissions, this),
+	campaignName(Campaign.name),		//missioncampaign.h globals
+	campaignDescr(Campaign.desc),
+	campaignTechReset(Campaign.flags & CF_CUSTOM_TECH_DATABASE)
+{
+	for (int i=0; i<Campaign.num_missions; i++) {
+		if (! missionData.contains(Campaign.missions[i].name)) {
+			mission temp{};
```

Comment:
I don't think the `{}` are needed if it's just invoking the default ctor. Or is this a style preference I was unaware of?

Reply:
It is a style choice one of mine, iirc I got a clang warning in one strange case and preferred to be explicit everywhere. But if it is irritating here / doesn't fit general code style, I'll remove them.

Reply:
Ah ok. Not sure if there's an SCP style guide, but if not, I'm OK with a bit of style variation.

The only reason it's significant here is that I'm legally blind, and reading even very large print by sight is super slow/difficult. Thus, having extra punctuation like this makes PR review a little harder for me.

FWIW I normally read code in VS using high-speed TTS, (screen reader) but for PR reviews, reading by sight tends to work better for understanding the changes.

Reply:
The general style guide is to use whatever is in the file.  If its a completely new file then the original author is free to choose their own style... as long as they're consistent within the rest of the file.

Reply:
As for the usage of the C++11 default initializer style... I'd say OK to using it on anything that isn't a standard data type such as `int` or `size_t`, for those instead I would highly recommend using the more traditional definition of using operator=.